Outputs files may be aggregated including information for all samples or provided per sample. Per-sample files will be prefixed with respective aliases and represented below as {{ alias }}.

| Title | File path | Description | Per sample or aggregated |
|-------|-----------|-------------|--------------------------|
| Workflow report | ./wf-aav-qc-report.html | Report for all samples | aggregated |
| Per file read stats | ./fastq_ingress_results/per-read-stats.tsv.gz | A TSV with per read stats for all samples. | aggregated |
| Combined reference sequence | ./combined_reference.fa.gz | Reference file containing all AAV plasmid and host genome sequences. | aggregated |
| Per read alignment info | ./{{ alias }}/{{ alias }}_bam_info.tsv | The result of `seqkit bam`. | per-sample |
| Alignment file | ./{{ alias }}/{{ alias }}_align.bam | The resulting BAM file from mapping reads to the combined reference. | per-sample |
| Alignment file index | ./{{ alias }}/{{ alias }}_align.bam.bai | The index of the resulting BAM file from mapping reads to the combined reference. | per-sample |
| AAV structure assignment | ./{{ alias }}/{{ alias }}_aav_per_read_info.tsv | AAV per read genome subtypes. | per-sample |
| Transgene plasmid consensus | ./{{ alias }}/{{ alias }}_transgene_plasmsid_consensus.fasta.gz | The transgene plasmid consensus sequence generated by medaka. | per-sample |
| Transgene plasmid variants | ./{{ alias }}/{{ alias }}_transgene_plasmid_variants.vcf.gz | The transgene plasmid variants file generated by medaka. | per-sample |
